lib/cn/kvset_split.c

```diff
@@ -0,0 +1,619 @@
+        hlog_union(work->hlog_right, hlog);
+
+        /* Add the source kblock to the purge list to be destroyed later */
+        blk_list_append(&result->blks_purge, kblk->kb_kblk.bk_blkid);
+    } else if (split_idx >= 0) { /* no overlap, append this kblock to the left kvset */
+        blk_list_append(&result->blks_left.kblks, ks->ks_kblks[split_idx].kb_kblk.bk_blkid);
+
+        kbr_read_hlog(&ks->ks_kblks[split_idx].kb_kblk_desc, &hlog);
+        hlog_union(work->hlog_left, hlog);
+    }
+
+    /* Add kblocks in [split_idx + 1, nkblks - 1] to the left kvset
+     */
+    for (i = split_idx + 1; i < ks->ks_st.kst_kblks; i++) {
+        blk_list_append(&result->blks_right.kblks, ks->ks_kblks[i].kb_kblk.bk_blkid);
+
+        kbr_read_hlog(&ks->ks_kblks[i].kb_kblk_desc, &hlog);
+        hlog_union(work->hlog_right, hlog);
+    }
+
+    return 0;
+}
+
+/**
+ * Return a split vblock index for the specified range of vblocks [start, end] by comparing
+ * the min/max keys stored in a vblock footer against the split key.
+ *
+ * Return values:
+ *     start - 1: if all vblocks belongs to the right kvset             // case 1
+ *     i and *overlap = false: left: [start, i], right [i + 1, end]     // case 2
+ *     i and *overlap = true:  left: [start, i], right [clone(i), end]  // case 3
+ */
+static int
+get_vblk_split_index(
+    struct kvset *ks,
+    int           start, /* inclusive */
+    int           end,   /* inclusive */
+    const void   *split_key,
+    uint32_t      split_klen,
+    bool         *overlap)
+{
+    int i;
+
+    *overlap = false;
+    assert(start <= end && end < kvset_get_num_vblocks(ks));
+
+    for (i = start; i <= end; i++) {
+        struct vblock_desc *vbd = kvset_get_nth_vblock_desc(ks, i);
+        const void *min_key = vbd->vbd_mblkdesc.map_base + vbd->vbd_min_koff;
+        const void *max_key = vbd->vbd_mblkdesc.map_base + vbd->vbd_max_koff;
+        int rc1;
+
+        rc1 = keycmp(split_key, split_klen, min_key, vbd->vbd_min_klen);
+        if (rc1 < 0) {
+            return i - 1; /* case 1, 2 */
+        } else {
+            int rc2 = keycmp(split_key, split_klen, max_key, vbd->vbd_max_klen);
+            if (rc2 < 0) {
+                *overlap = true;
+                return i; /* case 2, 3 */
+            }
+        }
+    }
+
+    return end; /* case 2 */
+}
+
+static merr_t
+kvset_split_vblocks(
+    struct kvset           *ks,
+    const struct key_obj   *split_kobj,
+    struct kvset_split_res *result)
+{
+    struct kvset_vgroup_map *vgmap_src = ks->ks_vgmap;
+    struct kvset_vgroup_map *vgmap_left = result->vgmap_left, *vgmap_right = result->vgmap_right;
+    int vbidx_left = 0, vbidx_right = 0; /* tracks vblock index for the left and right kvsets */
+    int vgidx_left = 0, vgidx_right = 0; /* tracks vgroup index for the left and right kvsets */
+    char split_key[HSE_KVS_KEY_LEN_MAX];
+    uint32_t split_klen, nvgroups = kvset_get_vgroups(ks);
+    int i;
```

Comment:
It's feels kinda dirty that you're pulling stuff out of an argument named "result"...
This decl block is a bit too noisy for my tastes, would rather not see more than one initialization per line.  But perhaps that's just me...

Reply:
I've refactored the declaration block.